pydantic_ai_slim/pydantic_ai/_tool_manager.py

```diff
@@ -120,6 +124,11 @@ async def _call_tool(self, call: ToolCallPart, allow_partial: bool, wrap_validat
             else:
                 args_dict = validator.validate_python(call.args or {}, allow_partial=pyd_allow_partial)
 
+            async with self._tool_calls_lock:
```

Comment:
This makes me wonder if we should also have a lock around `self.failed_tools.add(name)` below... Did you run into issues without this?